keps/sig-apps/2599-minreadyseconds-for-statefulsets/README.md

```diff
@@ -0,0 +1,616 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The main goal of this enhancement is to introduce minReadySeconds as an 
+optional field to StatefulSets. 
+
+## Motivation
+
+minReadySeconds specifies the minimum number of seconds for which a newly 
+created Pod should be ready without any of its containers crashing, for it 
+to be considered available. This defaults to 0 (the pod will be considered 
+available as soon as it is ready).
+
+minReadySeconds is already available as an optional field for Deployments,
+DaemonSets. Enabling this option helps in bringing StatefulSets on par 
+with other workload controllers.
+
+### Goals
+
+StatefulSet Controller honoring minReadySeconds and mark Pod ready only if Pod 
+is available for given time mentioned in minReadySeconds.
+
+### Non-Goals
+
+Moving minReadySeconds to Pod Spec is beyond the scope of this KEP 
+because of following reasons:
+- The effort to change pod spec would be large.
+- Currently our workload controllers are inconsistent and we prioritize consistency of experience. 
+- StatefulSets are just different enough from daemonsets and deployments that real world use of minReadySeconds 
+  for stateful sets might influence any future design or point in a more appropriate direction (completeness).
+
+## Proposal
+
+StatefulSet spec should be expanded with an optional field called 
+minReadySeconds. To reflect the StatefulSet pods that honored minReadySeconds,
+an additional field called `AvailableReplicas` will also be added to
+StatefulSet status.
+
+### User Stories (Optional)
+
+#### Story 1
+As a Kubernetes user, I want to ensure that my StatefulSet Pods have the 
+desired state ready before they start serving requests. Some applications 
+should only be considered as available only when the state transfer has 
+happened properly.
+
+#### Story 2
+As a Kubernetes user, I want to leverage StatefulSet's minReadySeconds behind
+service loadbalancers in clouds. This is to prevent killing pods in rotation
+before new pods show up.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+We are proposing a new field called `minReadySeconds` whose default value will 
+be 0. In this mode, StatefulSet will behave exactly like its current behavior.
+Its possible we introduce a bug in the implementation. The mitigation 
```

Comment:
Be more specific about the risk of a bug, namely that we either:
- mark a pod as ready too soon
- mark as ready too late. 
and what that might potentially mean.